### ISSUE TITLE
Оценочка

### ISSUE THREAD

**Issue by Lerostre** [tadgeislamins/NLP#1]
В принципе всё прекрасно, разве что в некоторых местах, где мне было не лень, я написал, как можно сделать на мой взгляд покрасивее, ну и мб чуть больше комментариев было бы неплохо, но здесь задача довольно простая, можно в целом и без них. Ещё было бы здорово, если бы мне не пришлось самому заново скачивать страницы, так я бы мог проверить, что у нас код даёт одно и то же. Ну и за слишком простой парсинг сказали слегка оштрафовать, поэтому

9.5

```
X_lemmatized = X.copy()
for i in range(len(X)):
    X_lemmatized.loc[i, 'review'] = lemmatize(X.loc[i, 'review'])
```

Здесь можно чуть быстрее сделать без цикла, ещё так будет более по-питоновски, через метод `apply`. Он применяет функцию в аргументе ко всему столбцу и возвращает уже трансформированный. Тут данных немного, в принципе не критично, но на большом объёме может помочь

```
X_lemmatized = X.copy()
X_lemmatized["review"] = X_lemmatized["review"].apply(lemmatize)
```

```
accuracy = 0
for i in range(len(y_lemmatized)):
    review = y_lemmatized.loc[i, 'review']
    y_pred = determine_review_tonality(review)
    y_true = y_lemmatized.loc[i, 'label']
    if y_pred == y_true:
        accuracy += 1
accuracy = accuracy / len(y_lemmatized)
accuracy
```

Вот тут снова знание `apply` сделает код куда компактнее. Можно не итерироваться по массиву, а работать сразу с векторами, в нампае же можно сравнивать их поэлементно, получится вектор True Positive, так сказать, его среднее даст аккураси

```
y_pred = y_lemmatized["review"].apply(determine_review_tonality)
(y_pred == y_lemmatized["label"]).mean()

# а ещё лучше просто украсть готовое
# from sklearn.metrics import accuracy_score
# accuracy_score(y_pred, y_true)
```

**Comment by Lerostre** [comment]
### По коду

Хорошо, но хранить и сравнивать можно супер удобно через датафрейм, где у тебя, например, колонки "токен", "разбор1", "разбор2" и так далее, это во-первых в одном месте. во вторвх это почти векторы, там оптимизированные быстрые операции, в-третьих он красивее, ну и можно легко отслеживать всякие наны, если у тебя нет разборов, дубликаты и так далее.       
Ну собственно ты к этому и приходишь, это хорошо, но я уже написал, пусть останется, единственное, что можно мёрджить несколько датафреймов по колонке - `pd.merge(df1, df2, on="token")`, он оставит только общее по дефолту, например по колонке токена, чтобы устранить разницу в токенизации

Теперь представим, что у тебя уже всё собрано в колонки, есть словарь, который переводит один тег в другой. Перевести колонку тегов в другой можно одним методом `map`, он всё сделает, не надо писать никаких функций.

### По заданиям

Никогда не семпли из тестовой выборки, чтобы оценить качество, либо фиксируй сид при рандоме- `np.random.seed(#любое число)`, либо запускай код на всех объектах. Иначе у тебя всегда будет разная выборка, ты никогда не поймёшь, какое у тебя на самом деле качество. Это не критичная ошибка, но надо обратить внимание *(-.5)*. Не знаю, как я в тот раз не заметил

9.5*0.8 = 7.6

**Comment by tadgeislamins** [comment]
Спасибо, только не до конца поняла про никогда не семпли из тестовой выборки. (Я скорее не семплила из тестовой выборки, а семплила, чтобы её создать, потому что в задании нужно было, чтобы она состояла из 20, кажется, элементов, но при этом не просили фиксировать одинаковое количество положительных и отрицательных, поэтому я решила засемплить, чтобы это было более рандомно и соответствовало условию). Я знаю, что если, например, все работают с одним и тем же датасетом и есть необходимость потом сравнивать качество, то часто предлагается зафиксировать сид при рандоме тестовой выборки, чтобы было правомерно сравнивать качество у разных людей / у разных моделей. Но поскольку здесь у каждого свой датасет я решила, что это не необходимо. А никакого «настоящего» качества, которое «на самом деле» ведь не существует и в чём-то, как я это понимаю, разная тестовая выборка при разных запусках даже более полезна, потому что получается такая кросс-валидация на минималках, если запускаешь код в процессе несколько раз.

**Comment by Lerostre** [comment]
А сэмпл же тебе не гарантирует, что соотношение классов будет такое же. Можно ресэмплить через библиотеки, типа `imblearn`, но это более продвинутые штуки. Ну и ты же свой датасет по итогу сравниваешь, на разных моделях, надо фиксировать выборку, чтобы было правомерно, всё верно, что не так? Ну вот так получилось, что других людей нет, надо сравнивать с собой. Либо тогда записывать куда-то результаты валидации, потому усреднять. Вообще лучше в личке обсудить, если тебе интересно

Не надо сэмплить выборку для качества, в общем, я запрещаю, если без фиксации сида. Кросс-валидация тоже обычно фиксируется, кстати
